spec/src/main/asciidoc/architecture.asciidoc

```diff
@@ -268,17 +650,349 @@ public PublisherBuilder<IncomingMessage, OutgoingMessage> processMessages() {
+* `MANUAL` - the user is responsible for the acknowledgement, so the Reactive Messaging implementation does not apply implicit acknowledgement
+* `PRE_PROCESSING` - the Reactive Messaging implementation acknowledges the message before the annotated method or processing is executed
+* `POST_PROCESSING` - the Reactive Messaging implementation acknowledges the message once:
+
+  1. the method or processing completes if the method does not emit data
+  2. when the emitted data is acknowledged
+
+Each method signatures has different defaults and accept a different set of strategy.
+If the `Acknowledgment` annotation is not set, the default policy is applied.
+
+The following table indicates the defaults and supported acknowledgement for each supported signature:
+
+[cols="2a,1,1",options="header"]
+|===
+|Signature
+|Default Acknoledgement Strategy
+|Supported Strategies
+
+|
+[source, java]
+----
+@Outgoing("name") 
+Publisher<Message<O>> method()
+----
+| None
+| 
+
+|
+[source, java]
+----
+@Outgoing("channel") 
+Publisher<O> method()
+----
+| None
+| 
+
+|
+[source, java]
+----
+@Outgoing("channel") 
+PublisherBuilder<Message<O>> method()
+----
+| None
+| 
+
+|
+[source, java]
+----
+@Outgoing("channel") 
+PublisherBuilder<O> method()
+----
+| None
+| 
+
+|
+[source, java]
+----
+@Outgoing("channel") 
+Message<O> method()
+----
+| None
+| 
 
-Messages may be wrapped in a `Message`, which can be used to supply metadata, and also allows messages to be acknowledged. The contract for acknowledging messages is anything that accepts a `Message` is required to acknowledge it. So, if the application receives an incoming message wrapped in `Message`, it is responsible for invoking `Message.ack()`, and if the application publish an outgoing message wrapped in `Message`, then the spec implementation is responsible for invoking `Message.ack()`.
+|
+[source, java]
+----
+@Outgoing("channel")
+O method()
+----
+| None
+| 
+
+|
+[source, java]
+----
+@Outgoing("channel")
+CompletionStage<Message<O>> method()
+----
+| None
+| 
+
+|
+[source, java]
+----
+@Outgoing("channel") 
+CompletionStage<O> method()
+----
+| None
+| 
+
+|
+[source,java]
+----
+@Incoming("channel")
+Subscriber<Message<I>> method()
+----
+| Post-Processing	
+| None, Pre-Processing, Post-Processing, Manual
+
+|
+[source,java]
+----
+@Incoming("channel")
+Subscriber<I> method()
+----
+| Post-Processing	
+| None, Pre-Processing, Post-Processing
+
+|
+[source,java]
+----
+@Incoming("channel") 
+SubscriberBuilder<Message<I>> method()
+----
+| Post-Processing	
+| None, Pre-Processing, Post-Processing, Manual
+
+
+|
+[source,java]
+----
+@Incoming("channel")
+SubscriberBuilder<I> method()
+----
+| Post-Processing	
+| None, Pre-Processing, Post-Processing
+
+|
+[source,java]
+----
+@Incoming("channel")
+void method(I payload)
+----
+| Post-Processing	
+| None, Pre-Processing, Post-Processing
+
+|
+[source,java]
+----
+@Incoming("channel")
+CompletionStage<?> method(Message<I> msg)
```

Comment:
For shapes that return CompletionStage, for Post-Processing does the ack occur when the CompletionStage is completed or when the method returns?

Reply:
when the `CompletionStage` is completed.

Reply:
actually, it's a bit more complicated, the ack happens when the emitted message is acknowledged, so definitely after the completion stage completion.